### ISSUE TITLE
[BUG] `top_k_mask_logits` hangs on certain inputs

### ISSUE THREAD

**Issue by clu0** [flashinfer-ai/flashinfer#1049]
`flashinfer.sampling.top_k_mask_logits` seems to hang on certain `logits` input tensors. For instance, the following code hangs for me for minutes without terminating:

```
import flashinfer
import torch

if __name__ == "__main__":
    data = torch.load("debug/tensors_2550.pt")
    logits: torch.Tensor = data["logits"]
    top_k: torch.Tensor = data["top_k"]
    top_k.fill_(1)
    print(f"{logits.shape=}, {top_k.shape=}")
    print(f"{logits=}, {top_k=}")
    print(f"{logits.max(dim=-1)=}")
    top_k_logits = flashinfer.sampling.top_k_mask_logits(logits, top_k=top_k)
    print(f"success, {top_k_logits=}")
```

It prints the following logs, before hanging on the `top_k_mask_logits` step:

```
logits.shape=torch.Size([206, 100352]), top_k.shape=torch.Size([206])
logits=tensor([[    -inf,     -inf,     -inf,  ...,     -inf,     -inf,     -inf],
        [    -inf,     -inf,     -inf,  ...,     -inf,     -inf,     -inf],
        [    -inf,     -inf,     -inf,  ...,     -inf,     -inf,     -inf],
        ...,
        [-25.0000, -24.8750, -25.0000,  ..., -25.0000, -25.0000, -25.1250],
        [ -7.5938,  -7.5938,  -7.5938,  ...,  -7.9062,  -7.9062,  -7.9062],
        [-14.0625, -14.0625, -14.0625,  ..., -13.8750, -13.9375, -13.9375]],
       device='cuda:0', dtype=torch.bfloat16), top_k=tensor([1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
        1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
        1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
        1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
        1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
        1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
        1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
        1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
        1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1], device='cuda:0',
       dtype=torch.int32)
logits.max(dim=-1)=torch.return_types.max(
values=tensor([-1.8625e+01, -3.1719e+00, -1.2625e+01, -1.9250e+01, -8.2500e+00,
        -2.0625e+01,  2.2000e+01, -2.1375e+01,  8.5938e-01,  5.3516e-01,
         4.2725e-02,  1.9250e+01, -5.5695e-04,  6.8750e-01,  1.6797e-01,
        -2.1973e-01,  2.0508e-01,  1.8625e+01, -3.0273e-02,  1.9141e-01,
         2.2875e+01, -9.5825e-03, -1.0547e-01,  1.9375e+01,  3.1500e+01,
         3.1375e+01, -4.9023e-01, -2.0801e-01,  2.3625e+01,  3.2750e+01,
         6.0156e-01,  1.0703e+00,  6.0547e-01,  1.1125e+01,  2.2625e+01,
         2.2750e+01,  1.0156e+00,  1.9125e+01, -2.2949e-01,  8.0469e-01,
         1.9375e+01,  2.3000e+01,  1.0438e+01,  3.1875e+01,  2.3125e+01,
        -4.1406e-01, -4.9805e-01,  3.2750e+01,  9.7500e+00,  9.8125e+00,
         1.0469e+00,  3.2250e+01,  3.1875e+01,  2.8320e-01, -2.7148e-01,
         6.9531e-01,  9.4375e+00, -1.6602e-01,  1.0703e+00,  8.9844e-01,
         4.9414e-01,  2.4875e+01,  9.5312e-01,  9.5625e+00,  9.5703e-01,
         1.9875e+01,  8.6875e+00,  1.0062e+01, -1.5137e-01,  1.8750e-01,
         3.3250e+01,  2.0375e+01, -2.0898e-01,  7.3047e-01,  2.3750e+01,
         3.3008e-01, -2.0801e-01,  1.8625e+01,  2.6758e-01,  3.7891e-01,
         2.0125e+01, -2.6172e-01,  3.5938e-01,  2.3375e+01,  3.3500e+01,
         2.4414e-01, -1.8652e-01,  3.4000e+01,  2.0125e+01,  4.3945e-01,
         3.3750e+01, -2.5391e-01, -3.1055e-01,  3.4250e+01,  2.3750e+01,
         6.5234e-01,  4.7852e-01,  1.9250e+01,  3.5750e+01, -2.7148e-01,
         9.7500e+00,  3.5500e+01,  2.6758e-01,  3.0664e-01,  9.8438e-01,
        -1.7090e-01,  7.5391e-01,  7.8125e-01,  2.0500e+01,  1.1328e+00,
         8.9453e-01,  3.1750e+01, -1.3184e-01,  4.0820e-01,  1.1250e+00,
         9.7266e-01,  1.9250e+01,  8.3203e-01,  2.9500e+01,  9.2578e-01,
         8.4375e+00,  2.6562e-01,  7.9375e+00,  1.8555e-01, -1.8359e-01,
        -3.9844e-01,  2.5250e+01,  2.3750e+01,  3.3750e+01,  8.1875e+00,
        -3.9453e-01,  2.4625e+01,  9.0625e+00,  4.5117e-01,  9.7500e+00,
        -3.4375e-01, -3.0859e-01,  2.5250e+01, -3.4375e-01,  3.1836e-01,
        -3.5742e-01,  2.5250e+01, -4.0625e-01, -2.8516e-01,  2.5625e+01,
         1.8359e-01,  3.8867e-01, -4.3945e-01, -3.4961e-01,  8.4375e+00,
        -4.3945e-01,  2.6000e+01, -3.8477e-01,  3.3500e+01, -3.8477e-01,
        -2.9297e-01, -5.1562e-01, -3.8867e-01, -3.2422e-01,  1.6797e-01,
        -3.2422e-01,  2.3828e-01, -4.2578e-01,  8.9375e+00,  2.7734e-01,
        -3.5156e-01,  1.0703e+00,  1.0156e+00, -4.0625e-01, -2.6562e-01,
         9.4375e+00,  9.3359e-01,  9.0000e+00,  9.9375e+00, -4.2578e-01,
         1.0234e+00,  2.1125e+01,  9.4375e+00,  8.8125e+00,  9.9375e+00,
         9.0625e+00,  4.8047e-01,  3.1445e-01,  8.1250e-01,  4.1211e-01,
         1.5156e+00,  1.1016e+00,  1.1016e+00,  1.2656e+00,  9.4375e+00,
         1.2188e+00,  1.2422e+00,  9.1875e+00,  3.2500e+01,  3.8867e-01,
         2.4875e+01,  1.2031e+00,  4.9805e-01,  2.0250e+01,  4.4141e-01,
         1.6992e-01,  8.0859e-01,  2.0781e+00,  3.2656e+00,  1.5250e+01,
         8.8750e+00], device='cuda:0', dtype=torch.bfloat16),
indices=tensor([100022, 100022, 100022, 100022, 100022, 100022,    402, 100022,    428,
           355,    428,      3,    351,    428,    428,    351,    428,      3,
           351,    428,    402,    351,    351,      3,   2397,   2397,    362,
           351,    402,   2397,    355,   1285,    355,    574,    402,    402,
          1285,      3,    351,    428,      3,    402,    574,   2397,    402,
           351,    362,   2397,    574,    574,   1285,   2397,   2397,    355,
           351,    428,    574,    351,   1285,   1285,    428,    402,   1285,
           574,   1285,      3,    574,    574,    351,    355,   2397,      3,
           351,    428,    402,    355,    351,      3,    428,    355,      3,
           351,    355,    402,   2397,    355,    351,   2397,      3,    355,
          2397,    351,    351,   2397,    402,    428,    355,      3,   2397,
           351,    574,   2397,    355,    355,   1285,    351,   1285,   1285,
             3,    355,   1285,   2397,    351,    355,   1285,   1285,      3,
          1285,   2397,   1285,    574,    355,    574,    428,    351,    362,
           402,    402,   2397,    574,    362,    402,    574,    355,    574,
           362,    362,    402,    362,    355,    362,    402,    362,    362,
           402,    428,    355,    362,    362,    574,    362,    402,    362,
          2397,    362,    362,    362,    362,    362,    355,    362,    428,
           362,    574,    355,    362,   1285,   1285,    362,    351,    574,
          1285,    574,    574,    362,   1285,      3,    574,    574,    574,
           574,    428,    428,    428,    355,    574,   1285,   1285,   1285,
           574,   1285,   1285,    574,   2397,    428,    402,   1285,    428,
             3,    352,    351,    354,    352,     67,    920,  21316],
       device='cuda:0'))
```

The same script still fails when I change the logit dtype to say `float32`.

This is running on Hopper, on flashinfer `'0.2.3+cu124torch2.5'`, torch version `'2.5.1+cu124'`, and with the following CUDA versions:
```
NVIDIA-SMI 550.90.07              Driver Version: 550.90.07      CUDA Version: 12.4 
```

**Comment by xslingcn** [comment]
Hi @clu0, could you please provide the tensor dump for reproduction? Thanks!

**Comment by clu0** [comment]
Had some trouble uploading the `.pt` file via github. Here is a google drive link: https://drive.google.com/file/d/1Rs9qkaU39eVQcKfcCCek94i-fdNmatyN/view?usp=sharing

You should be able to retrieve the tensors with my script above
```
    logits: torch.Tensor = data["logits"]
    top_k: torch.Tensor = data["top_k"]
```

**Comment by clu0** [comment]
It is possible that the algorithm on this specific input is just really slow. I tried sorting the array with triton and that took minutes.

I initially thought it was a bug because `top_k_mask_logits` returns very quickly when the logits is of the same shape (`(206, 100352)`) but filled with random values. In fact triton also fails to sort the random valued tensor in a reasonable time.

**Comment by yzh119** [comment]
@clu0 the renorm algorithm should return fast for any input, so there might be some bug with the code, we are investigating the issue.

**Comment by xslingcn** [comment]
For now the issue is due to the `-inf`s in the inputs, when initializing the loop bounds it will cause both of our pivots set to `-inf` so loop won't stop. We'll need to decide how we want to handle such inputs.

cc @yzh119

**Comment by yzh119** [comment]
Per @xslingcn 's investigation, your input tensor has `-inf` values which our kernel don't handle.

We will create a PR to handle these abnormal values.

**Comment by yzh119** [comment]
Should have been fixed in #1050 .

**Comment by clu0** [comment]
Sounds good, thanks for the fix!
